internal/datastore/memdb/caveat.go

```diff
@@ -0,0 +1,118 @@
+package memdb
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+
+	"github.com/hashicorp/go-memdb"
+)
+
+const tableCaveats = "caveats"
+
+type caveat struct {
+	id         datastore.CaveatID
+	name       string
+	expression []byte
+}
+
+func (c *caveat) CoreCaveat() *core.Caveat {
+	return &core.Caveat{
+		Name:       c.name,
+		Expression: c.expression,
+	}
+}
+
+func (r *memdbReader) ReadCaveatByName(name string) (*core.Caveat, error) {
+	r.lockOrPanic()
+	defer r.Unlock()
+
+	tx, err := r.txSource()
+	if err != nil {
+		return nil, err
+	}
+	return r.readCaveatByName(tx, name)
+}
+
+func (r *memdbReader) ReadCaveatByID(ID datastore.CaveatID) (*core.Caveat, error) {
+	r.lockOrPanic()
+	defer r.Unlock()
+
+	tx, err := r.txSource()
+	if err != nil {
+		return nil, err
+	}
+	return r.readCaveatByID(tx, ID)
+}
+
+func (r *memdbReader) readCaveatByID(tx *memdb.Txn, ID datastore.CaveatID) (*core.Caveat, error) {
+	found, err := tx.First(tableCaveats, indexID, ID)
+	if err != nil {
+		return nil, err
+	}
+	if found == nil {
+		return nil, fmt.Errorf("caveat with id %d not found: %w", ID, datastore.ErrCaveatNotFound)
+	}
+	c := found.(*caveat)
+	return c.CoreCaveat(), nil
+}
+
+func (r *memdbReader) readCaveatByName(tx *memdb.Txn, name string) (*core.Caveat, error) {
+	found, err := tx.First(tableCaveats, indexName, name)
+	if err != nil {
+		return nil, err
+	}
+	if found == nil {
+		return nil, fmt.Errorf("caveat with name %s not found: %w", name, datastore.ErrCaveatNotFound)
+	}
+	c := found.(*caveat)
+	return c.CoreCaveat(), nil
+}
+
+func (rwt *memdbReadWriteTx) WriteCaveats(caveats []*core.Caveat) ([]datastore.CaveatID, error) {
+	rwt.lockOrPanic()
+	defer rwt.Unlock()
+	tx, err := rwt.txSource()
+	if err != nil {
+		return nil, err
+	}
+	return rwt.writeCaveat(tx, caveats)
+}
+
+func (rwt *memdbReadWriteTx) writeCaveat(tx *memdb.Txn, caveats []*core.Caveat) ([]datastore.CaveatID, error) {
+	ids := make([]datastore.CaveatID, 0, len(caveats))
```

Comment:
Are we handling revisions for caveats in here? They will need to be revisioned the same as namespaces, which I don't think are revisioned in memdb? @jakedt am I correct?

Reply:
As far as I can tell namespaces are not revisioned in MemDB

Reply:
Everything is revisioned in memdb by keeping snapshots of the database after each mutation.

Reply:
https://github.com/authzed/spicedb/blob/main/internal/datastore/memdb/memdb.go#L224

Reply:
ah, my bad! I was somehow looking for some explicit app level MVCC. MemDB handles this for us, thanks for the pointer Jake!

Reply:
added a test to demonstrate snapshot reads for caveats: https://github.com/authzed/spicedb/pull/807/commits/b22266d54f0506b68c4a8290c98eeae7f5c7a71b

in order to support this I had to turn the caveat-write operation into an upsert